### PR TITLE
[router]: Fix incorrect routing sorting for static paths

### LINKED ISSUE #30879: [3.5.21] - [id] wildcard matched higher than explicit selector
### Minimal reproducible example

https://github.com/austin43/expo-router-bug-3.5.21

### Which package manager are you using? (Yarn is recommended)

bun

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

metro

### Summary

After installing 3.5.21, I noticed our explicit routes were no longer being matched over [id] wildcard routes. When we routed to "/nested/create" for example, it would render the "/nested/[id]" route with create as the id param.

Click either button on the homepage of example app, and notice how each goes to the [id] page, rather than one going to [id] and the other to create.

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.0 - ~/.volta/tools/image/node/18.16.0/bin/node
      Yarn: 1.22.19 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
      npm: 9.5.1 - ~/.volta/tools/image/node/18.16.0/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
      Android SDK:
        API Levels: 28, 31
        Build Tools: 30.0.2, 30.0.3, 31.0.0
    IDEs:
      Android Studio: 2021.3 AI-213.7172.25.2113.9123335
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.25 => 51.0.25 
      expo-router: ~3.5.21 => 3.5.21 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.74.5 => 0.74.5 
      react-native-web: ~0.19.10 => 0.19.12 
    Expo Workflow: managed

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

Fix incorrect route sorting for static paths. #30266 changed route sorting to prioritise routes within the same group. However, it failed to take into account that static routes outside the current group should still be ranked higher. This PR address that issue.

I also observed that PR also fixes staying within the same group for some hoisted route scenarios (fixed by changing how to generate `expandedRouteNames`).

Fix: [#30879](https://github.com/expo/expo/issues/30879)

# How

- Fix `expandedRouteNames` to always expand the names of hoisted routes
- Ensure the sorting function checks for static vs dynamic before group prioritisation

**Note**

Layout files have their own route config, as you are able to route to them and they use `initialRouteName` to determine the actual state. The current logic always puts these configs ranked lower than route files - but if we just changed if a config is dynamic or not it would rank layout configs incorrectly. Therefor they are excluded via `config.hasChildren` and should continue to be ranked lower than route configs.

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/__tests__/navigation.test.ios.tsx [hunk 1]
```diff
@@ -1325,3 +1325,55 @@ describe('stack unwinding', () => {
     expect(router.canGoBack()).toBe(true); //
   });
 });
+
+it.only('should always prefer static routes over dynamic ones', async () => {
```
Comment (kitten) [inline packages/expo-router/src/__tests__/navigation.test.ios.tsx:1329]:
```suggestion
it('should always prefer static routes over dynamic ones', async () => {
```

remaining `.only`

File: packages/expo-router/src/fork/getStateFromPath.ts [hunk 2]
```diff
@@ -671,9 +708,9 @@ const createConfigItem = (
     path,
     // The routeNames array is mutated, so copy it to keep the current state
     routeNames: [...routeNames],
-    expandedRouteNames: screen.includes('/')
-      ? [...routeNames.slice(0, -1), ...screen.split('/')]
-      : [...routeNames],
+    expandedRouteNames: routeNames.flatMap((name) => {
+      return name.split('/');
```
Comment (kitten) [inline packages/expo-router/src/fork/getStateFromPath.ts:712]: note: I can't immediately tell from the code whether the normalization from before is redundant

**Review by kitten — APPROVED** [review]

**Comment by timoisalive** [comment]
Hi! Is this fix supposed to be in the latest version 3.5.23, or is it still unpublished as the changelog states?

This issue is breaking our app, and would be great to get the fix in as soon as possible... Thanks!

**Comment by DavidAmyot** [comment]
> Hi! Is this fix supposed to be in the latest version 3.5.23, or is it still unpublished as the changelog states?
> 
> This issue is breaking our app, and would be great to get the fix in as soon as possible... Thanks!

I have the same issue and it is indeed breaking our app too. @marklawlor From what I can see, this seems to be in the unpublished changelogs, but can you please confirm whether or not it is already in 3.5.23 and if not, maybe a small ETA of when it'll be published please?

Thank you!

**Comment by DavidAmyot** [comment]
@timoisalive I just cloned the expo repo and replaced my `node_modules/expo-router` folder with the expo-router folder from the cloned expo folder inside of _packages/expo-router_ and the issue is not present for me anymore. 

For me, the issue was also causing trouble even just to navigate to `/_sitemap` since it would try to render the first alphabetical route in the group I was in. For example, if I was in `app/(app)/(ccav)/assigments-list` and would try to navigate with a Link whose href was `/_sitemap`, it would return my `app/(app)/(ccav)/[job]` route instead of the sitemap. 

The real issue for us was when we forced 403 or 401 global redirection such as token expiration and I'd try to redirect to `app/(app)/units-list` and everything would break.

So yeah I don't think this pr is in 3.5.23 (in the changelog it shows in unpublished), hope it'll be released soon though.

**Comment by statico** [comment]
This PR is not in 3.5.23 as we're experiencing this issue with 3.5.23.

**Comment by DavidAmyot** [comment]
> This PR is not in 3.5.23 as we're experiencing this issue with 3.5.23.

Yeah, it's not unfortunately. As I said in my previous comment, a temporary fix for us was to clone the **expo** repo and create a repo out of the **packages/expo-router** folder from the **expo** repo and use the github url of the repo as the package in our package.json.

**Comment by smrubin** [comment]
Is this fix in a published expo-router version yet? I don't believe its in 3.5.24 either.

**Pull request merged**